tsdb/exemplar.go

```diff
@@ -0,0 +1,229 @@
+	// Return exemplars as a new slice from oldest to newest.
+	for i := el.oldest; i < len(el.list); i++ {
+		ret = append(ret, el.list[i])
+	}
+	for i := 0; i < el.oldest; i++ {
+		ret = append(ret, el.list[i])
+	}
+	return ret
+}
+
+// NewExemplarStorage creates new in-memory storage for exemplars.
+func NewInMemExemplarStorage(len int) *InMemExemplarStorage {
+	return &InMemExemplarStorage{
+		exemplars: make(map[string]*exemplarList),
+		len:       len,
+	}
+}
+
+func (es *InMemExemplarStorage) Appender() storage.ExemplarAppender {
+	return es
+}
+
+// TODO: separate wrapper struct for queries?
+func (es *InMemExemplarStorage) Querier(ctx context.Context) (storage.ExemplarQuerier, error) {
+	return es, nil
+}
+
+// Select returns exemplars for a given set of series labels hash.
+func (es *InMemExemplarStorage) Select(l labels.Labels) ([]exemplar.Exemplar, error) {
+	if _, ok := es.exemplars[l.String()]; !ok {
+		return nil, nil
+	}
+	return es.exemplars[l.String()].sorted(), nil
+}
+
+func (es *InMemExemplarStorage) AddExemplar(l labels.Labels, t int64, e exemplar.Exemplar) error {
+	// todo: if we're doing time brackets for exemplars, ie 1min, 15min, 30min, 60min etc
+	// check if t should bump an existing exemplar out of the storage?
+
+	// Ensure no empty labels have gotten through.
+	l = l.WithoutEmpty()
+
+	if _, ok := es.exemplars[l.String()]; !ok {
+		es.exemplars[l.String()] = newExemplarList(es.len)
+	}
+	return es.exemplars[l.String()].add(e)
+}
+
+// For use in tests, clears the entire exemplar storage
+func (es *InMemExemplarStorage) Reset() {
+	es.exemplars = make(map[string]*exemplarList)
+}
+
+// ***************************************************************
+// single circular buffer for all exemplars
+type CircularExemplarStorage struct {
+	lock      sync.RWMutex
+	index     map[string]int
+	exemplars []circularBufferEntry
+	nextIndex int
+	len       int
+}
+
+type circularBufferEntry struct {
+	exemplar     exemplar.Exemplar
+	seriesLabels labels.Labels // need to store labels so we can double check when querying
+	prev         int           // index of previous exemplar in circular for the same series
+}
+
+// If we assume the average case 95 bytes per exemplar we can fit 5651272 exemplars in
+// 1GB of extra memory, accounting for the fact that this is heap allocated space.
+func NewCircularExemplarStorage(len int) *CircularExemplarStorage {
+	return &CircularExemplarStorage{
+		exemplars: make([]circularBufferEntry, len),
+		index:     make(map[string]int),
+		len:       len,
+	}
+}
+
+func (ce *CircularExemplarStorage) Appender() storage.ExemplarAppender {
+	return ce
+}
+
+// TODO: separate wrapper struct for queries?
+func (ce *CircularExemplarStorage) Querier(ctx context.Context) (storage.ExemplarQuerier, error) {
+	return ce, nil
+}
+
+// Select returns exemplars for a given set of series labels hash.
+func (ce *CircularExemplarStorage) Select(l labels.Labels) ([]exemplar.Exemplar, error) {
```

Comment:
I was quite surprised select does not return iterator, but I guess it's fine for now

Reply:
I can't think of a reason to not return an iterator, that was just more work for an MVP when I was working on this like 6 months ago :)